gomock/controller_114_test.go

```diff
@@ -0,0 +1,132 @@
+// Copyright 2020 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+// +build go1.14
+
+package gomock_test
+
+import (
+	"testing"
+
+	"github.com/golang/mock/gomock"
+)
+
+func (e *ErrorReporter) Cleanup(f func()) {
+	e.t.Helper()
+	e.t.Cleanup(f)
+}
+
+func TestMultipleDefers(t *testing.T) {
+	reporter := NewErrorReporter(t)
+	reporter.Cleanup(func() {
+		reporter.assertLogf("In Go 1.14+ you no longer need to `ctrl.Finish()` if a *testing.T is passed to `NewController(...)`")
+	})
+	ctrl := gomock.NewController(reporter)
+	ctrl.Finish()
+}
+
+// Equivalent to the TestNoRecordedCallsForAReceiver, but without explicitly
+// calling Finish.
+func TestDeferNotNeededFail(t *testing.T) {
+	reporter := NewErrorReporter(t)
+	subject := new(Subject)
+	var ctrl *gomock.Controller
+	reporter.Cleanup(func() {
+		reporter.assertFatal(func() {
+			ctrl.Call(subject, "NotRecordedMethod", "argument")
+		}, "Unexpected call to", "there are no expected calls of the method \"NotRecordedMethod\" for that receiver")
+	})
+	ctrl = gomock.NewController(reporter)
+}
+
+func TestDeferNotNeededPass(t *testing.T) {
+	reporter := NewErrorReporter(t)
+	subject := new(Subject)
+	var ctrl *gomock.Controller
+	reporter.Cleanup(func() {
+		reporter.assertPass("Expected method call made.")
+	})
+	ctrl = gomock.NewController(reporter)
+	ctrl.RecordCall(subject, "FooMethod", "argument")
+	ctrl.Call(subject, "FooMethod", "argument")
+}
+
+func TestOrderedCallsInCorrect(t *testing.T) {
```

Comment:
With a couple of these tests I had to get a little fancy. Their implementation relied on Finish not being called to verify things. Cleanup acts in a LIFO matter. By registering a cleanup before the one in NewController, I can assert things after Finish() has been called.